### PR TITLE
Start Sequentially Minting at `_startTokenId()` instead of hardcoded 0

### PR DESCRIPTION
**Pull request by Pczek**
As described in #45  NFTs with token ID `0` can cause some headaches for developers down the line. 

I renamed the `currentIndex` to `_nextTokenId` as this describes its use better as well it's an internal variable. But let me know your thoughts about it.
During development I stumbled over this function declaration 
`function getOwnershipAt(uint256 index)` in `ERC721AExplicitOwnershipMock.sol`
For me it was a bit misleading. I expected `_ownership` to be an multidimensional array containing the history of the ownership of an NFT. However, it's a flat mapping between basically the `tokenId` to the `TokenOwnership`. I would suggested renaming the `index` parameter to `tokenId`.
https://github.com/chiru-labs/ERC721A/blob/main/contracts/mocks/ERC721AExplicitOwnershipMock.sol#L19-L21

### CODE REVIEW

**Comment by MrMcGoats** [comment]
You should probably add `.vscode` to the git ignore

**Comment by Pczek** [comment]
@MrMcGoats Done. Thanks for pointing that out.

I've added new tests as well. One for `tokenOfOwnerByIndex` is currently failing. 
Any suggestions how to adapt it?

https://github.com/chiru-labs/ERC721A/blob/main/contracts/ERC721A.sol#L85-L108

**Comment by Pczek** [comment]
Fixed the `tokenOfOwnerByIndex` function now as well. I wrote some test for the. The PR is ready for review, happy to get some feedback and merge it if approved. 👍

**Comment by jpegdigital** [comment]
Hi, thought of one more edge case with _exists(). I think `return tokenId < currentIndex;` fails when checking tokenId 0 or the last tokenId? Maybe something like this to fix `return tokenId > 0 && tokenId <= currentIndex;`

I'm surprised there aren't existing tests already that fail for the boundaries. Haven't had a chance to run the test suite though.

**Comment by Vectorized** [comment]
This should have the flexibility to allow users to choose between zero and one indexed tokenIds, with minimal additional gas cost and minimal API change.

File: contracts/ERC721A.sol
Comment (Vectorized) [inline contracts/ERC721A.sol:70]: Can be unchecked.

File: contracts/ERC721A.sol
Comment (Vectorized) [inline contracts/ERC721A.sol:78]: Can be unchecked.

File: contracts/ERC721A.sol
Comment (Vectorized) [inline contracts/ERC721A.sol:63]: Should not be hardcoded. Users should have the flexibility to choose if they want to start from 0 or 1, minimally.

I've written a version that uses a function override at https://gist.github.com/Vectorized/e6f47b0c1de008bb5983852224e41f69.

File: contracts/extensions/ERC721AOwnersExplicit.sol
Comment (Vectorized) [inline contracts/extensions/ERC721AOwnersExplicit.sol:16]: Can be changed to `!=` to save tiny gas https://github.com/chiru-labs/ERC721A/pull/59. `_nextTokenId` won't go below 1 anyway.

File: contracts/extensions/ERC721AOwnersExplicit.sol
Comment (Vectorized) [inline contracts/extensions/ERC721AOwnersExplicit.sol:17]: Should init nextOwnerToExplicitlySet in constructor.

Comment (Vectorized) [inline contracts/extensions/ERC721AOwnersExplicit.sol:17]: Should check and set the _nextOwnerToExplicitlySet local variable instead for gas savings. 

See https://github.com/chiru-labs/ERC721A/pull/59

**Comment by Pczek** [comment]
@Vectorized Thanks for the review. The change requests check out, will incorporate them probably today. Not sure what you mean by `Can be unchecked.`? I guess it relates to the option to make the starting index flexible?
I wonder if you think it should start at 0 or 1 or if even other starting indexes should be valid as well. 

@jpegdigital good call, I've added some `_exists` tests and adapted the function my changes yesterday already. ✌️ 
https://github.com/chiru-labs/ERC721A/pull/66/files#diff-833552e5bd4a4be3ada89b7874eedc6762fb91ed0890dda22e82a1936e6d60d8R280-R282

**Comment by Vectorized** [comment]
> @Vectorized Thanks for the review. The change requests check out, will incorporate them probably today. Not sure what you mean by `Can be unchecked.`? I guess it relates to the option to make the starting index flexible?

Wrap those lines in `unchecked {}` blocks for gas savings, since they cannot overflow and underflow.

See: https://gist.github.com/Vectorized/e6f47b0c1de008bb5983852224e41f69

**Comment by Pczek** [comment]
@Vectorized thanks for the gist, it helped a lot. I've adapted my fork to your suggestions. Works fine now. It needed minor adjustments. 
I've brought back the original test files and plainly copied them for new mock contracts which uses the override to set the starting token id to one. Not sure how I feel about the copy-pasted code, not sure how to do it more elegantly atm.

File: contracts/ERC721A.sol
Comment (Vectorized) [inline contracts/ERC721A.sol:105]: It may be more gas efficient to compare it to a local copy of `_nextTokenId`.

```
uint256 end = _nextTokenId;
...
unchecked {
             for (uint256 i = _startTokenId(); i < end; i++) {
```

**Comment by Vectorized** [comment]
@Pczek 

Thanks for your hard work. I was actually preparing to work on this, until I saw your PR. 
Don't worry about copying. Just want to contribute. 

We rise together.
We build together.
We grow together.

**Comment by TimTinkers** [comment]
Likewise, thanks @Pczek.

**Comment by ottogutierrez** [comment]
Thanks @Pczek! I'm excited to see this one on the next release :)

File: contracts/ERC721A.sol
Comment (Pczek) [reply contracts/ERC721A.sol:123]: Here are the results in gas for this change. I can only spot a difference in the gas used during deployment. At least it's saving. But seems only to be the case if another `_startTokenId` than `0` is used. 

I've committed this change. 👍 

# Before with `numMintedSoFar`
Screenshot 2022-02-10 at 10 00 44


# After with `end`
Screenshot 2022-02-10 at 10 02 36

**Comment by Pczek** [comment]
@Vectorized @chiru-labs Who can do the final review and approve it to be merged? 
IMHO the PR is ready to be merged.

**Comment by cygaar** [comment]
Once we push the burn function in, I'll review this one :)

Thanks for all the hard work and bearing with us

**Comment by Vectorized** [comment]
@Pczek I've opened a PR to your fork with the added burn functionality. https://github.com/Pczek/ERC721A/pull/1

**Comment by Pczek** [comment]
@Vectorized @cygaar merged https://github.com/Pczek/ERC721A/pull/1 ✅ 
All tests succeed ✅

**Comment by cygaar** [comment]
Awesome thanks, this is the next PR on the docket

**Comment by Pczek** [comment]
@cygaar that would be great! I've just resolved the latest conflict and adapted the code of mine to work with the recent changes of the `main` branch.

Regarding a gas report. Here's a screenshot of the current `main` and my fork side by side. The only difference I'm able to spot is the cost of deployment. However, these are expected as more code is deployed. 🤷 


Screenshot 2022-02-18 at 14 03 48

**Comment by Vectorized** [comment]
Interesting how the mint actually used lower gas on average when the starting index is 1 instead of 0. 

When the index starts at 0, the very first mint will change `_currentIndex` from 0 to 1, which costs more gas (20000 vs 5000).

**Comment by ahbanavi** [comment]
@Pczek I've opened a PR https://github.com/Pczek/ERC721A/pull/3 to your branch for adding `_totalMinted` function (#124).

**Comment by cygaar** [comment]
Thanks for your PR @ahbanavi let's wait for that to get merged and then I'll do an in-depth review of this PR

**Comment by Pczek** [comment]
@Vectorized @ahbanavi just merged your PR https://github.com/Pczek/ERC721A/pull/3

Test seem to run fine, I haven't adapted anything else since the PR of @ahbanavi .

Anything left to do? I saw you opened up a new PR @Vectorized ?

**Comment by Vectorized** [comment]
@Pczek Welcome back! 

I'll close that PR first. Let's wait for @cygaar.

**Comment by cygaar** [comment]
Going to review this today

File: contracts/mocks/ERC721AStartOneMock.sol
Comment (cygaar) [inline contracts/mocks/ERC721AStartOneMock.sol:8]: You should just inherit from `ERC721AMock` instead of redeclaring everything again

File: contracts/mocks/ERC721AStartOneExplicitOwnershipMock.sol
Comment (cygaar) [inline contracts/mocks/ERC721AStartOneExplicitOwnershipMock.sol:8]: Inherit from `ERC721AOwnersExplicitMock` instead of duplicating everything

File: test/ERC721AStartOne.test.js
Comment (cygaar) [inline test/ERC721AStartOne.test.js:10]: I think a better approach to our tests would be parameterize this class variable to change between the 0-index and 1-index contracts so we don't have to duplicate most of the code.

You would then have specific tests for `ownerOf` and tokenId. You would probably need a separate context for 1-index specific test cases.

I want to minimize the work we need to do anytime we add/change some functionality in our contract.

File: test/ERC721A.test.js
Comment (cygaar) [inline test/ERC721A.test.js:22]: Can you add totalMinted test cases for ERC721ABurnable?

**Review by cygaar — CHANGES_REQUESTED** [review]
The actual code looks fine, @Vectorized I think we can make the readme/comment changes in another PR. I'm more concerned about test quality and reducing copy-pasted code

File: test/ERC721A.test.js
Comment (ahbanavi) [reply test/ERC721A.test.js:22]: I add one here:  
https://github.com/chiru-labs/ERC721A/blob/f0a5bb42f85300e9857932c5ddb79164dd07b124/test/extensions/ERC721ABurnable.test.js#L56-L63  
Are there any other situations?

Comment (Vectorized) [reply test/ERC721A.test.js:22]: This is hard. 

I'm thinking of simply adding a `directBurn` function to `ERC721AMock` if we want to test `_totalMinted` with `_burn` and `_startTokenId`. 

The route of having a `ERC721ABurnableStartOne`, or making `ERC721AMock` subclass `ERC721ABurnable` feels uncomfortable.

File: test/ERC721AStartOne.test.js
Comment (Pczek) [reply test/ERC721AStartOne.test.js:10]: Agree, to remove most of the duplicate test code. This is a good hint, I will look into that. When I copied them I didn't know better. Will try to adapt it today.

**Comment by Pczek** [comment]
@Vectorized @cygaar please check my recent changes to make a parameterized test. How do you like it this way?
If this is fine for you we can adapt the rest of the duplicated tests for the ExplicitOwnership extension as well.

[`2ba810d` (#66)](https://github.com/chiru-labs/ERC721A/pull/66/commits/2ba810d4c241b36b6c82bca122c65a6ba699ae6c#diff-e166966f732548cffaf8c83a2732c01fdf82cad2d861478e0b1dd43ba00d9fe2)

File: contracts/mocks/ERC721AStartTokenIdMock.sol
Comment (Vectorized) [inline contracts/mocks/ERC721AStartTokenIdMock.sol:9]: Make `startTokenId` immutable for zero-cost abstraction. Otherwise, each call to _startTokenId will incur gas. 

You can also make this public and remove the `getStartTokenId` function.

Consider moving this class to a separate file, since you are going to subclass it from `OwnersExplicit` too.

**Comment by cygaar** [comment]
LGTM. @Pczek parameterize the ownership explicit test and this is good to go. Appreciate all the work you and everyone has put in to get this out

File: test/ERC721A.test.js
Comment (Pczek) [reply test/ERC721A.test.js]: Would be cool to test these features individually, however, these are quite connected. 
So I created a new mock combining both in one of my recent commits:
[`e16a79a` (#66)](https://github.com/chiru-labs/ERC721A/pull/66/commits/e16a79a5bc8382aef740cc843e3c4d04a3ff03d9)

**Comment by Pczek** [comment]
Alright, I've parameterized now the following tests:
`ERC721A.test`
`ERC721AOwnersExplicit.test`
`ERC721ABurnable.test`

All of them now also run their test suite for a contract which is modified with the `StartTokenIdHelper.sol` contract to set the override the `_startTokenId()` function dynamically.

Afaik no further tasks are open on this PR until another review or it's ready to be merged.

CC @Vectorized @cygaar

File: test/ERC721A.test.js
Comment (cygaar) [inline test/ERC721A.test.js:355]: Styling nit: can you add a new line between lines 354 and 355?

**Review by cygaar — APPROVED** [review]

File: test/ERC721A.test.js
Comment (Pczek) [reply test/ERC721A.test.js:357]: Sure done ✅  Good catch 👍

**Pull request merged**
